### ISSUE TITLE
bad torch.float32 accuracy

### ISSUE THREAD

**Issue by sun7-she** [pytorch/pytorch#18426]
### Please note that this issue tracker is not a help form and this issue will be closed.

We have a set of [listed resources available on the website](https://pytorch.org/resources). Our primary means of support is our discussion forum:

- [Discussion Forum](https://discuss.pytorch.org/)

**Comment by sun7-she** [comment]
run the code1: 

`import torch
import numpy as np

n=4
c=3
h=32
w=32

inputs = torch.FloatTensor(n,c,h,w).fill_(0)
inputs = np.asarray(inputs)
inputs_sum = 0.0

for j in range(c):
    for k in range(h):
        for p in range(w):
            inputs[0,j,k,p] = (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4
            inputs_sum += (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4

for i in range(n):
    print('inputs[%d].sum: %.6f' % (i, inputs[0:i+1,:,:,:].sum()))

print("%.6f %.6f" % (inputs.sum(), inputs_sum))
`

we get the outputs:
inputs[0].sum: 39782.402344
inputs[1].sum: 39782.402344
inputs[2].sum: 39782.398438
inputs[3].sum: 39782.398438
39782.398438 39782.400000


run the code2:

`import torch
import numpy as np

n=1
c=3
h=32
w=32

inputs_t = np.zeros([n,c,h,w])
inputs_n = np.zeros([n,c,h,w])
inputs_tensor_from_numpy = torch.from_numpy(inputs_n)
inputs_numpy_from_tensor = np.asarray(inputs_t)
inputs_tensor = torch.FloatTensor(n,c,h,w).fill_(0)
inputs_tensor64 = torch.zeros([n,c,h,w], dtype=torch.float64)
inputs_numpy = np.zeros([n,c,h,w])
inputs_sum_num = 0.0

for j in range(c):
    for k in range(h):
        for p in range(w):
            inputs_tensor_from_numpy[0,j,k,p] = (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4
            inputs_numpy_from_tensor[0,j,k,p] = (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4
            inputs_tensor[0,j,k,p] = (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4
            inputs_tensor64[0,j,k,p] = (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4
            inputs_numpy[0,j,k,p] = (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4
            inputs_sum_num += (0+1)*1 + (j+1)*0.2 + (k+1)*0.3 + (p+1)*0.4

inputs_sum_tensor_from_numpy = 0.0
inputs_sum_numpy_from_tensor = 0.0
inputs_sum_tensor = 0.0
inputs_sum_tensor64 = 0.0
inputs_sum_numpy = 0

for j in range(c):
    for k in range(h):
        for p in range(w):
            inputs_sum_tensor_from_numpy += inputs_tensor_from_numpy[0,j,k,p]
            inputs_sum_numpy_from_tensor += inputs_numpy_from_tensor[0,j,k,p]
            inputs_sum_tensor += inputs_tensor[0,j,k,p]
            inputs_sum_tensor64 += inputs_tensor64[0,j,k,p]
            inputs_sum_numpy += inputs_numpy[0,j,k,p]

print('tensor from numpy',inputs_sum_tensor_from_numpy)
print('numpy from tensor',inputs_sum_numpy_from_tensor)
print('tensor',inputs_sum_tensor,inputs_sum_tensor.dtype)
print('tensor64',inputs_sum_tensor64)
print('numpy',inputs_sum_numpy,inputs_sum_numpy.dtype)
print('num',inputs_sum_num)
`

we get the outputs:

tensor from numpy tensor(39782.4000, dtype=torch.float64)
numpy from tensor 39782.399999999994
tensor tensor(39782.4023) torch.float32
tensor64 tensor(39782.4000, dtype=torch.float64)
numpy 39782.399999999994 float64
num 39782.399999999994

The acc of float32 tensor is not good.........-.-

**Comment by albanD** [comment]
Duplicate of #18427
